src/lib/support/jsontlv/JsonToTlv.cpp

```diff
@@ -0,0 +1,344 @@
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <lib/support/Base64.h>
+#include <lib/support/SafeInt.h>
+#include <lib/support/jsontlv/JsonToTlv.h>
+#include <lib/support/jsontlv/JsonTypes.h>
+#include <nlohmann/json.hpp>
+#include <sstream>
+
+using json = nlohmann::ordered_json;
+
+namespace chip {
+
+namespace {
+
+std::vector<std::string> splitString(const std::string & input, char separator)
+{
+    std::vector<std::string> substrings;
+    std::stringstream ss(input);
+    std::string substring;
+
+    while (std::getline(ss, substring, separator))
+    {
+        substrings.push_back(substring);
+    }
+
+    return substrings;
+}
+
+CHIP_ERROR JsonTypeStrToTlvType(const char * elementType, TLV::TLVType & type, bool & isDouble)
+{
+    if (strcmp(elementType, kElementTypeInt) == 0)
+    {
+        type = TLV::kTLVType_SignedInteger;
+    }
+    else if (strcmp(elementType, kElementTypeUInt) == 0)
+    {
+        type = TLV::kTLVType_UnsignedInteger;
+    }
+    else if (strcmp(elementType, kElementTypeBool) == 0)
+    {
+        type = TLV::kTLVType_Boolean;
+    }
+    else if (strcmp(elementType, kElementTypeFloat) == 0)
+    {
+        type     = TLV::kTLVType_FloatingPointNumber;
+        isDouble = false;
+    }
+    else if (strcmp(elementType, kElementTypeDouble) == 0)
+    {
+        type     = TLV::kTLVType_FloatingPointNumber;
+        isDouble = true;
+    }
+    else if (strcmp(elementType, kElementTypeBytes) == 0)
+    {
+        type = TLV::kTLVType_ByteString;
+    }
+    else if (strcmp(elementType, kElementTypeString) == 0)
+    {
+        type = TLV::kTLVType_UTF8String;
+    }
+    else if (strcmp(elementType, kElementTypeNull) == 0)
+    {
+        type = TLV::kTLVType_Null;
+    }
+    else if (strcmp(elementType, kElementTypeStruct) == 0)
+    {
+        type = TLV::kTLVType_Structure;
+    }
+    else if (strncmp(elementType, kElementTypeArray, strlen(kElementTypeArray)) == 0)
+    {
+        type = TLV::kTLVType_Array;
+    }
+    else
+    {
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+bool isUnsignedInteger(const std::string & s)
+{
+    size_t len = s.length();
+    if (len == 0)
+    {
+        return false;
+    }
+    for (size_t i = 0; i < len; i++)
+    {
+        if (!isdigit(s[i]))
+        {
+            return false;
+        }
+    }
+    return true;
+}
+
+bool isSignedInteger(const std::string & s)
+{
+    if (s.length() == 0)
+    {
+        return false;
+    }
+    if (s[0] == '-')
+    {
+        return isUnsignedInteger(s.substr(1));
+    }
+    return isUnsignedInteger(s);
+}
+
+CHIP_ERROR ParseJsonKey(const std::string & key, TLV::Tag & tag, TLV::TLVType & type, bool & isDouble, TLV::TLVType & subType,
```

Comment:
The pairing of `TLV::TLVType, isDouble` seems quite an annoying leakage of abstraction. How about creating a separate container that contains both the tlv type and the information if it is double or not? That way it is clearer what type is connected to what value and we can also ensure that isDouble is never set/available when TLV type is not floating point.

Reply:
updated. Introduced `struct` that combines both values: {TLVType, isDouble}